pkg/s3utils/utils.go

```diff
@@ -181,3 +182,126 @@ func EncodePath(pathName string) string {
 	}
 	return encodedPathname
 }
+
+// Validator interface for writing two types of validators, one
+//
+// - Check for common validation of any input string argument.
+// - CheckStrict for common validation + some stricter guidelines.
+//
+// The errors and requirements for the interface are left to
+// user to decide.
+type Validator interface {
+	// Common validation function.
+	Check(s string) error
+
+	// Common validation + stricter guidelines function.
+	CheckStrict(s string) error
+}
+
+// BucketNameValidator validates bucket names and satisfies
+// the Validator interface of Check functions.
+type BucketNameValidator struct{}
+
+// NewBUcketNameValidator - initializes a new bucket name validator.
+func NewBucketNameValidator() Validator {
+	return &BucketNameValidator{}
+}
+
+// Common checker for both stricter and basic validation.
+func (s *BucketNameValidator) checkCommon(bucketName string, strict bool) (err error) {
+	if strings.TrimSpace(bucketName) == "" {
+		return errors.New("Bucket name cannot be empty")
+	}
+	if len(bucketName) < 3 {
+		return errors.New("Bucket name cannot be smaller than 3 characters")
+	}
+	if len(bucketName) > 63 {
+		return errors.New("Bucket name cannot be greater than 63 characters")
+	}
+	// Split on dot and check each piece conforms to rules.
+	allNumbers := true
+	pieces := strings.Split(bucketName, ".")
+	for _, piece := range pieces {
+		if len(piece) == 0 || piece[0] == '-' ||
+			piece[len(piece)-1] == '-' {
+			// Current piece has 0-length or starts or
+			// ends with a hyphen.
+			return errors.New("Bucket name contains invalid characters")
+		}
+		// Now only need to check if each piece is a valid
+		// 'label' in AWS terminology and if the bucket looks
+		// like an IP address.
+		isNotNumber := false
+		for i := 0; i < len(piece); i++ {
+			switch {
+			case (piece[i] >= 'A' && piece[i] <= 'Z' ||
+				piece[i] == '-'):
+				if strict && piece[i] != '-' {
+					// For all capitalized characters return error in
+					// strict mode.
+					return errors.New("Bucket name contains invalid characters")
+				}
+				// Found a non-digit character, so
+				// this piece is not a number.
+				isNotNumber = true
+			case (piece[i] >= 'a' && piece[i] <= 'z' ||
+				piece[i] == '-'):
+				// Found a non-digit character, so
+				// this piece is not a number.
+				isNotNumber = true
+			case piece[i] >= '0' && piece[i] <= '9':
+				// Nothing to do.
+			default:
+				return errors.New("Bucket name contains invalid characters")
+			}
+		}
+		allNumbers = allNumbers && !isNotNumber
+	}
+	// Does the bucket name look like an IP address?
+	if len(pieces) == 4 && allNumbers {
+		return errors.New("Bucket name cannot be an ip address")
+	}
```

Comment:
Well this is copied from server code @krishnasrinivas . We can keep the regex no big deal i think.